--- ./configure.ac.orig	2018-04-23 22:17:33.000000000 +0800
+++ ./configure.ac	2018-06-20 13:20:23.998591900 +0800
@@ -2498,7 +2498,7 @@ fi
 
 ### Use -lrsvg-2 if available, unless '--with-rsvg=no' is specified.
 HAVE_RSVG=no
-if test "${HAVE_X11}" = "yes" || test "${HAVE_NS}" = "yes" || test "${opsys}" = "mingw32"; then
+if test "${HAVE_X11}" = "yes" || test "${HAVE_NS}" = "yes" || test "${HAVE_W32}" = "yes"; then
   if test "${with_rsvg}" != "no"; then
     RSVG_REQUIRED=2.11.0
     RSVG_MODULE="librsvg-2.0 >= $RSVG_REQUIRED"
--- ./src/image.c.orig	2017-02-03 19:25:45.000000000 +0900
+++ ./src/image.c	2017-04-22 12:06:14.183459000 +0900
@@ -9907,7 +9907,7 @@
 #if defined (HAVE_RSVG)
   DEFSYM (Qsvg, "svg");
   ADD_IMAGE_TYPE (Qsvg);
-#ifdef HAVE_NTGUI
+#if defined HAVE_NTGUI && !defined CYGWIN
   /* Other libraries used directly by svg code.  */
   DEFSYM (Qgdk_pixbuf, "gdk-pixbuf");
   DEFSYM (Qglib, "glib");
